File: solutions/go/src/playground/algorithm/permutations.go
package algorithm

func permute(nums []int) [][]int {
	var result [][]int
	permuteR(nums, 0, &result)
	return result
}

func permuteR(nums []int, start int, result *[][]int) {
	if start == len(nums)-1 {
		*result = append(*result, nums)
		return
	}

	for i := start; i < len(nums); i++ {
		nums[start], nums[i] = nums[i], nums[start]
		tmp := make([]int, len(nums))
		copy(tmp, nums)
		permuteR(tmp, start+1, result)
		nums[start], nums[i] = nums[i], nums[start]
	}
}

File: solutions/java/src/main/java/playground/algorithm/Permutations.java
package playground.algorithm;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import playground.util.Utils;

public class Permutations {

    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        permute(nums, 0, result);
        return result;
    }

    private void permute(int[] nums, int start, List<List<Integer>> list) {
        if (start == nums.length - 1) {
            list.add(Arrays.stream(nums).boxed().collect(Collectors.toList()));
            return;
        }

        for (int i = start; i < nums.length; i++) {
            Utils.swap(nums, i, start);
            permute(nums, start + 1, list);
            Utils.swap(nums, i, start);
        }
    }
}
